### ISSUE TITLE
ALB 502 failures 

### ISSUE THREAD

**Issue by prakashbalaji** [grpc/grpc-java#11597]
Hello, 

We are using spring grpc and deploy it in EKS and expose it via ALB load balancer. When we run our performance benchmark tests using k6 and connecting to pod directly, we have all the calls successful. But when we run the same tests pointing to Load balancer there are always 400-500 calls failing due to ELB 502. We see the below exception in the stack trace. We would appreciate any pointers on debugging this issue, we are mainly looking for guidance if any with respect to running grpc services behind load balancers. We have tweaked lot of timeouts on both grpc and AWS ALB end and we keep getting this exception. 

```
grpc.server.enable-keep-alive=true
grpc.server.keep-alive-time=120s
grpc.server.keep-alive-timeout=120s
grpc.server.permit-keep-alive-without-calls=true
```

<img width="861" alt="Screenshot 2024-10-03 at 11 10 33 PM (1)" src="https://github.com/user-attachments/assets/8d284b54-6163-43b6-badd-424688e2e3d7">
<img width="862" alt="Screenshot 2024-10-03 at 11 17 28 PM (1)" src="https://github.com/user-attachments/assets/3f737533-a84f-4fcf-b610-c284580b3f02">

io.grpc.StatusRuntimeException: INTERNAL: Encountered end-of-stream mid-frame at io.grpc.Status.asRuntimeException(Status.java:525) at io.grpc.internal.AbstractServerStream$TransportState.deframerClosed(AbstractServerStream.java:234) at io.grpc.netty.NettyServerStream$TransportState.deframerClosed(NettyServerStream.java:153)

Thanks

**Comment by ejona86** [comment]
You mention ELB 502 but the error doesn't mention that at all. I don't know what that means. The end-of-stream mid-frame means the proxy is at least partly broken; it probably was trying to handle an error and is closing the stream instead of resetting the stream, but those are different semantically.

You might check if there are request timeouts close to 15 seconds. But the proxy is the only thing that probably knows what is wrong, so it would need to give you a clue.

**Comment by prakashbalaji** [comment]
@ejona86 -Thanks for taking a look at the issue, The exception attached is from our target pod and below is the access logs for one such exception. 

The elb response code is 502. 
The target response code is - 
The -1 is in the response time and it has a value of -1 and based on AWS doc -1 indicates
`This value is set to -1 if the load balancer doesn't receive a response from a target. This can happen if the target closes the connection before the idle timeout or if the client sends a malformed request.`

The syntax of the rest of the message is based on the table [here](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html)

Attaching some tcpdump details as well from our packet capture data to see if it aids in any direction you can give us to investigate. 

```
grpcs 2024-10-03T05:04:18.590574Z app/table-store/81bc15c44c51e313 10.3.42.106:64337 10.6.44.123:9092 0.000 0.003 -1 502 - 215 96 "POST https://<MASK_URL>/table_store.service.TableStore/FetchData HTTP/2.0" "grpc-go/1.64.1" TLS_AES_128_GCM_SHA256 TLSv1.3 arn:aws:elasticloadbalancing:us-west-2:<AWS_ACCOUNT>:targetgroup/<TARGET_GROUP> "Root=1-66fe25d2-6ee07cad0b694e38126d1fb3" "<MASK_URL>" "arn:aws:acm:us-west-2:<AWS_ACCOUNT>:certificate/<SOME HASH>" 0 2024-10-03T05:04:18.586000Z "forward" "-" "-" "10.6.44.123:9092" "-" "-" "-" TID_5624d29f8fc60143a2440752eeffbb68
```

<img width="1728" alt="Screenshot 2024-10-04 at 10 40 17 PM" src="https://github.com/user-attachments/assets/16ddb71b-8f69-426e-b2f1-44e2c74afca6">

**Comment by ejona86** [comment]
The TCP disconnect is from a 10 second timer on the LB after the last bytes sent from the backend. Since the minimum direct-to-pod latency was ~200 ms, and the main communication happens within ~1 ms, then most likely the response hadn't been sent and the request timed out.

The gRPC keepalive stuff is pretty similar to TCP keepalive, which means it is pretty different than HTTP keeepalive.

I think you are looking for a defaults-to-10-seconds timeout in ALB, and then configure it with a larger value.

**Comment by prakashbalaji** [comment]
@ejona86 - Thanks for the pointers, based on some more anlalysis of tcpdump this is the pattern we could establish, there are FIN packets from load balancer to target in 30 seconds interval and all the 502 calls in our ALB Proxy are the ones where the FIN is overlapping at that time. 

None of our settings are in 30seconds range and feel like some default is kicking in, looking at the logs and screenshot below, can you point to any of the settings that we have to review? 

<img width="448" alt="Screenshot 2024-10-07 at 11 32 20 AM" src="https://github.com/user-attachments/assets/c6799ea5-bda8-401f-a061-55f2210209d9">

![Screenshot 2024-10-07 at 9 38 26 AM](https://github.com/user-attachments/assets/416bc7ab-5301-4b9e-ae88-07b8e3266485)
![Screenshot 2024-10-07 at 9 44 30 AM](https://github.com/user-attachments/assets/5eb2543c-c046-48f0-aa90-840718903e71)

**Comment by prakashbalaji** [comment]
Thanks @ejona86  for your response to this issue. We managed to find the underlying issue and would like to flag it here just incase it helps someone else.

We had a simple grpc request of one string param as the request parameter. For all the successful calls we received these bytes to target via load balancer.
```
000000001d0a1b66645f72657461696c6572735f6575726f70655f636f756e747279 
000000001e0a1c66645f72657461696c6572735f6e6f6e5f6575726f70655f696e746c 
00000000150a1366645f72657461696c6572735f6575726f7065 
000000000e0a0c66645f72657461696c657273 
```

When the requests failed we received a slight variation to the above valid strings though we tested with the same 4 strings. 
```
00012e01040000002d8386449d61238e82c4849ec2ae82d9dcc42afbc71d05dc7279
00012d01040000003b8386449d61238e82c4849ec2ae82d9dcc42afbc71d05dc93746c
00012e0104000000358386449d61238e82c4849ec2ae82d9dcc42afbc71d05dc7279
00012e0104000000318386449d61238e82c4849ec2ae82d9dcc42afbc71d05dc7279
```

This corruption of data led us to this thread - https://github.com/grpc/grpc-java/issues/11284 and we find it strange that we did not get to this comment earlier as it has the exact same issue. 
https://github.com/grpc/grpc-java/issues/11284#issuecomment-2168113850

Upgrading the grpc-java from 1.61 to 1.66 fixed the underlying corruption issue and the calls via ALB are all successful.
